### PR TITLE
Add a CDL Hold Frontend

### LINKED ISSUE #4091: When a resource is unavailable for charging, display a button to a user to let them place a hold in the queue.
## Implementation Notes

Add `CDL::CDLController#hold`, which uses `ChargeManager#create_hold`

The button should be added to `app/viewer/cdl_checkout.html.erb`.

It should say "Create Reservation" and be accompanied by some text which says "This material is currently checked out. However, if you create a reservation we will notify you when it's available for check out and hold it for you for up to 1 hour."

Also add a note like `There are <x> people currently awaiting this item.`

### PR DESCRIPTION
**Pull request by tpendragon**
Closes #4091 

## Reservation Creation Screen

Screen Shot 2020-07-16 at 12 25 00 PM

## Reservation Waiting Screen

Screen Shot 2020-07-16 at 12 25 11 PM

## Screen when reservation is active.

Screen Shot 2020-07-16 at 12 26 53 PM

### CODE REVIEW

File: app/resources/cdl/cdl_controller.rb [hunk 1]
```diff
@@ -17,6 +17,16 @@ def charge
       redirect_to auth_viewer_path(params[:id])
     end
 
+    def hold
+      return forbidden unless current_user
+      @charge_manager = charge_manager(params[:id])
+      @charge_manager.create_hold(netid: current_user.uid)
+      redirect_to auth_viewer_path(params[:id])
+    rescue CDL::HoldExists
+      flash[:alert] = "You already have a reservation for this item."
```
Comment (christinach) [inline app/resources/cdl/cdl_controller.rb:26]: can we change this to: 'You have already reserved this item.'

File: app/views/viewer/cdl_checkout.html.erb [hunk 2]
```diff
@@ -12,8 +12,20 @@
...
 
   <p>This institution reserves the right to refuse to accept a copying order if, in its judgment, fulfillment of the order would involve violation of copyright law.</p>
-  <%= button_to "Check Out for 3 Hours", { action: "charge", controller: "cdl/cdl" }, class: "btn btn-primary", disabled: !@charge_manager.available_for_charge?(netid: current_user&.uid), params: { id: params[:id] } %>
-  <% if !@charge_manager.available_for_charge?(netid: current_user&.uid) %>
-    <b>This item is currently checked out. The estimated wait time is <%= @charge_manager.estimated_wait_time %>.</b>
+  <% if @charge_manager.available_for_charge?(netid: current_user&.uid) %>
+    <%= button_to "Check Out for 3 Hours", { action: "charge", controller: "cdl/cdl" }, class: "btn btn-primary", params: { id: params[:id] } %>
+  <% elsif @charge_manager.hold?(netid: current_user&.uid) %>
+    <%= button_to "Create Reservation", { action: "hold", controller: "cdl/cdl" }, class: "btn btn-primary", params: { id: params[:id] }, disabled: true %>
+    <p>
```
Comment (christinach) [inline app/views/viewer/cdl_checkout.html.erb:19]: can we change the caption to: 'Reserve'

File: app/views/viewer/cdl_checkout.html.erb [hunk 3]
```diff
@@ -12,8 +12,20 @@
...
+    There are <%= @charge_manager.hold_index(netid: current_user&.uid)%> reservation(s) ahead of you. We will send you an email notification when your reservation is ready for check out.
+    </p>
+  <% else %>
+    <%= button_to "Create Reservation", { action: "hold", controller: "cdl/cdl" }, class: "btn btn-primary", params: { id: params[:id] } %>
+    <p>
```
Comment (christinach) [inline app/views/viewer/cdl_checkout.html.erb:24]: can we change the caption to: 'Reserve'

File: app/views/viewer/cdl_checkout.html.erb [hunk 4]
```diff
@@ -12,8 +12,20 @@
...
+    This item is currently checked out. However, if you create a reservation we will notify you when it is available for check out and hold it for you from that point for up to 1 hour.
+    </p>
```
Comment (christinach) [inline app/views/viewer/cdl_checkout.html.erb:26]: Is it necessary to have the phrase: `from that point`?

**Comment by tpendragon** [comment]
@christinach Made the requested text changes.

**Pull request merged**